File: part1/jlox/Lox.java
package Lox;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

enum TokenType {
    // Single-character tokens
    LEFT_PAREN, RIGHT_PAREN, LEFT_BRACE, RIGHT_BRACE,
    COMMA, DOT, MINUS, PLUS, SEMICOLON, SLASH, STAR,

    // One or two character tokens
    BANG, BANG_EQUAL,
    EQUAL, EQUAL_EQUAL,
    GREATER, GREATER_EQUAL,
    LESS, LESS_EQUAL,

    // Literals
    IDENTIFIER, STRING, NUMBER,

    // Keywords
    AND, CLASS, ELSE, FALSE, FUN, IF, NIL, OR,
    PRINT, RETURN, SUPER, THIS, TRUE, VAR, WHILE,

    EOF
}

class Token{
    final TokenType type;
    final String lexeme;
    final Object literal;
    final int line;

    Token(TokenType type, String lexeme, Object literal, int line){
        this.type = type;
        this.lexeme = lexeme;
        this.literal = literal;
        this.line = line;
    }

    public String toString(){
        return type + " " + lexeme + " " + literal;
    }
}

public class Lox {

  static boolean hadError = false;

  public static void main(String[] args) throws IOException {
    if (args.length > 1) {
      System.out.println("Usage: jlox [script]");
      System.exit(64); 
    } else if (args.length == 1) {
      runFile(args[0]);
    } else {
      System.out.println("Running REPL");
      runPrompt();
    }
  }

  static void runFile(String arg){
      byte[] bytes = Files.readAllBytes(Paths.get(path));
      run(new String(bytes, Charset.defaultCharset()));
      if(hadError) System.exit(65);
  }

  /**
   * Runs the Lox REPL
   */
  static void runPrompt(){
      InputStreamReader input = new InputStreamReader(System.in);
      BufferedReader reader = new BufferedReader(input);

      for(;;){
          System.out.print("> ");
          String line = reader.readLine();
          if (line == null) break;
          run(line);
          hadError = false;
      }
  }

  private static void run(String source){
      Scanner scanner = new Scanner(source);
      List<Token> tokens = scanner.scanTokens();

      // TODO
      for(Token token: tokens){
          System.out.println(token);
      }
  }

  static void error(int line, String message){
      report(line, "", message);
  }

  private static void report(int line, String where, String message){
      System.err.println(
          "[line " + line + "] Error" + where + ": " + message
      );
      hadError = true;
  }
}

File: part1/lox.py
import sys
from scanner import Scanner
from util import Token, TokenType
from parser import Parser, ParserException
from interpreter import Interpreter
from runtime_error import RuntimeError
from tool.ast_printer import AstPrinter


class Lox:
    had_error: bool = False
    had_runtime_error: bool = False

    @classmethod
    def main(cls, *args):
        if len(args) > 1:
            print(f"Usage: {sys.argv[0]} [script]")
            sys.exit(64)
        elif len(args) == 1:
            cls.run_file(args[0])
        else:
            print("Running REPL")
            cls.run_prompt()

    @classmethod
    def run_file(cls, arg: str):
        with open(arg, "rb") as f:
            bytes = f.read()

        cls.run(bytes.decode("utf-8"))
        if cls.had_error:
            sys.exit(65)
        if cls.had_runtime_error:
            sys.exit(70)

    @classmethod
    def run_prompt(cls):
        while True:
            line = input("> ")
            if len(line.strip()) == 0:
                break
            cls.run(line)

    @classmethod
    def run(cls, source: str):
        try:
            scanner = Scanner(source, cls.error)
            tokens = scanner.scan_tokens()
            parser = Parser(tokens)
            statements = parser.parse()

            # for debugging (now OUTDATED)
            # for statement in statements:
            #     print(AstPrinter().print(statement))

            interpreter = Interpreter()
            interpreter.interpret(statements)

        except ParserException as parse_exception:
            cls.error(parse_exception.token, parse_exception.message)
        except RuntimeError as runtime_error:
            cls.runtime_error(runtime_error)

        if cls.had_error:
            return

    @classmethod
    def error(cls, line: int | Token, message: str):
        if isinstance(line, int):
            cls.report(line, "", message)
        else:
            if line.type == TokenType.EOF:
                cls.report(line.line, " at end", message)
            else:
                cls.report(line.line, f" at '{line.lexeme}'", message)

    @classmethod
    def runtime_error(cls, error: RuntimeError):
        print(error.message + f"\n[line {error.token.line}] ")
        cls.had_runtime_error = True

    @classmethod
    def report(cls, line: int, where: str, message: str):
        s = f"[line {line}] Error {where}: {message}"
        print(s)
        cls.had_error = True
        return s
